pkg/client/client.go

```diff
@@ -0,0 +1,410 @@
+	"error": "action params failed to be un-marshaled",
+})
+
+// ActionErrInvalidParams is returned to Elastic Agent as result to an action request
+// when the request params are invalid for the action.
+var ActionErrInvalidParams = utils.JSONMustMarshal(map[string]string{
+	"error": "action params invalid",
+})
+
+// ActionErrUnmarshableResult is returned to Elastic Agent as result to an action request
+// when the action was performed but the response could not be marshalled to send back to
+// the agent.
+var ActionErrUnmarshableResult = utils.JSONMustMarshal(map[string]string{
+	"error": "action result failed to be marshaled",
+})
+
+// Action is an action the client exposed to the Elastic Agent.
+type Action interface {
+	// Name of the action.
+	Name() string
+
+	// Execute performs the action.
+	Execute(map[string]interface{}) (map[string]interface{}, error)
+}
+
+// StateInterface defines how to handle config and stop requests.
+type StateInterface interface {
+	// OnConfig is called when the Elastic Agent is requesting that the configuration
+	// be set to the provided new value.
+	OnConfig(string)
+
+	// OnStop is called when the Elastic Agent is requesting the application to stop.
+	OnStop()
+
+	// OnError is called when an errors occurs communicating with Elastic Agent.
+	//
+	// These error messages are not given by the Elastic Agent, they are just errors exposed
+	// from the client-side GRPC connection.
+	OnError(error)
+}
+
+// Client manages the state and communication to the Elastic Agent.
+type Client struct {
+	target          string
+	opts            []grpc.DialOption
+	token           string
+	impl            StateInterface
+	actions         map[string]Action
+	cfgIdx          uint64
+	cfg             string
+	expected        proto.StateExpected_State
+	observed        proto.StateObserved_Status
+	observedMessage string
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	client proto.ElasticAgentClient
+	cfgLock   sync.RWMutex
+	obsLock   sync.RWMutex
+
+	// overridden in tests to make fast
+	minCheckTimeout time.Duration
+}
+
+// New creates a client connection to Elastic Agent.
+func New(target string, token string, impl StateInterface, actions []Action, opts ...grpc.DialOption) *Client {
+	actionMap := map[string]Action{}
+	if actions != nil {
+		for _, act := range actions {
+			actionMap[act.Name()] = act
+		}
+	}
+	return &Client{
+		target:          target,
+		opts:            opts,
+		token:           token,
+		impl:            impl,
+		actions:         actionMap,
+		cfgIdx:          0, // 0 is initial, no config state
+		expected:        proto.StateExpected_RUNNING,
+		observed:        proto.StateObserved_STARTING,
+		observedMessage: "Starting",
+		minCheckTimeout: CheckinMinimumTimeout,
+	}
+}
+
+// Start starts the connection to Elastic Agent.
+func (c *Client) Start(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	c.ctx = ctx
```

Comment:
Nit: Why not assign it directly on line 121? Same for cancel?

Reply:
Good point, fixed.